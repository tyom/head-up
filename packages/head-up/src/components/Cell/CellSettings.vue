<template>
  <div class="CellSettings">
    <h2>Settings</h2>
  </div>
</template>

<style scoped>
.CellSettings {
  padding: 1rem;
  position: absolute;
  background-color: #000e;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}
</style>
