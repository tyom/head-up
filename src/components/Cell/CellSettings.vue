<template>
  <div class="CellSettings">
    <h2>Settings</h2>
  </div>
</template>

<style scoped>
.CellSettings {
  @apply absolute inset-0 p-3 bg-gray-800 text-white text-lg;
}
</style>
