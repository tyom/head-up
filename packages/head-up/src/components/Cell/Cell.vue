<template>
  <div class="Cell">
    <header v-if="isEditable || title" class="header">
      <h2 class="title">{{ title }}</h2>
      <button
        v-if="isEditable"
        :class="{_toggled: showSettings}"
        class="header-button"
        @click.stop="handleSettingsClick"
      >
        <v-icon name="ellipsis-h"/>
      </button>
    </header>
    <div class="container">
      <div :class="bodyClass" class="body">
        <slot v-if="$slots.default"/>
        <template v-else>
          <component
            v-for="(item, idx) in content"
            :is="item.type"
            :key="idx"
            v-bind="item.props"
            class="body-item"
          />
        </template>
      </div>
      <transition name="revealSettings">
        <CellSettings v-if="isEditable && showSettings"/>
      </transition>
    </div>
  </div>
</template>

<script>
import CellSettings from './CellSettings';

export default {
  name: 'Cell',
  components: {
    CellSettings,
  },
  props: {
    title: {
      type: String,
      default: '',
    },
    content: {
      type: Array,
      default: () => [],
    },
    padded: {
      type: Boolean,
      default: false,
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      showSettings: false,
    };
  },
  computed: {
    isEditable() {
      return this.editable && this.isEditing();
    },
    bodyClass() {
      const content = this.$slots.default ? this.$slots.default : this.content;
      return {
        [`u-grid-${content.length}-x`]: !!content.length,
        padded: this.padded,
      };
    },
  },
  inject: ['isEditing'],
  methods: {
    handleSettingsClick() {
      this.showSettings = !this.showSettings;
    },
  },
};
</script>

<style scoped>
.Cell {
  border: 1px solid #0005;
  transition: 0.3s;
  display: flex;
  flex-direction: column;
  user-select: none;

  &:hover {
    transition: 0.1s;
  }
}

.header {
  font-size: 0.55em;
  color: #fff6;
  position: relative;
  text-transform: uppercase;
  background-color: #0003;
  padding: 0.7em 1em;
  line-height: 1;
  justify-content: center;
  display: flex;
  align-items: center;
}

.header-button {
  border: 0;
  padding: 0.5em 1em;
  background: none;
  color: #fff;
  position: absolute;
  z-index: 1;
  top: 0;
  right: 0;
  bottom: 0;
  opacity: 0.7;
  outline: 0;

  &:hover {
    opacity: 1;
  }

  &:focus-visible {
    background-color: rgba(#20416d, 0.7);
    box-shadow: 0 0 0 1px #fff;
  }

  &._toggled {
    opacity: 1;
    background-color: #20416d;
  }

  &:active {
    opacity: 1;
    transition: none;
    background-color: #20416d;
  }
}

.title {
  font-size: inherit;
}

.container {
  position: relative;
  flex: 1;
  display: flex;
}

.body {
  grid-gap: 1px;
  background-color: rgba(#000, 0.2);
  flex: 1;
  justify-content: stretch;
  align-items: stretch;
}

.body > * {
  background-color: rgba(#fff, 0.03);
}

.padded > * {
  padding: 0.5em;
}

.revealSettings-enter-active,
.revealSettings-leave-active {
  transition: 0.3s;
}
.revealSettings-enter,
.revealSettings-leave-to {
  opacity: 0;
}
</style>
