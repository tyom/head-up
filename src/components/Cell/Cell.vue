<template>
  <div class="Cell">
    <header v-if="isEditable || title" class="header">
      <h2 class="title">{{ title }}</h2>
      <button
        v-if="isEditable"
        :class="{ _toggled: showSettings }"
        class="header-button"
        @click.stop="handleSettingsClick"
      >
        <v-icon name="ellipsis-h" />
      </button>
    </header>
    <div class="body-wrapper">
      <div :class="bodyClass" class="body" :style="bodyStyle">
        <slot v-if="$slots.default" />
        <template v-else>
          <component
            v-for="(item, idx) in content"
            :is="item.type"
            :key="idx"
            v-bind="item.props"
            class="body-item"
          />
        </template>
      </div>
      <transition name="revealSettings">
        <CellSettings v-if="isEditable && showSettings" />
      </transition>
    </div>
  </div>
</template>

<script>
import merge from 'lodash/merge';
import store from '../../store';
import { createGridTemplate } from '../../utils';
import CellSettings from './CellSettings';

export default {
  name: 'Cell',
  components: {
    CellSettings,
  },
  props: {
    grid: String,
    title: {
      type: String,
      default: '',
    },
    content: {
      type: Array,
      default: () => [],
    },
    padded: {
      type: Boolean,
      default: false,
    },
    editable: {
      type: Boolean,
      default: false,
    },
  },
  created() {
    if (!this.$slots.default) {
      return;
    }
    const { areas } = createGridTemplate(this.grid);
    this.$slots.default.forEach((vnode, idx) => {
      vnode.data = merge(vnode.data, {
        style: {
          gridArea: areas[idx],
        },
      });
    });
  },
  data() {
    return {
      showSettings: false,
    };
  },
  computed: {
    isEditable() {
      return this.editable && store.state.editMode;
    },
    bodyStyle() {
      return {
        gridTemplateAreas: createGridTemplate(this.grid).template,
      };
    },
    bodyClass() {
      return {
        padded: this.padded,
      };
    },
  },
  methods: {
    handleSettingsClick() {
      this.showSettings = !this.showSettings;
    },
  },
};
</script>

<style scoped>
.Cell {
  @apply flex flex-col flex-grow transition duration-300 select-none p-px antialiased;
  background-color: #6277a244;

  &:hover {
    @apply duration-100;
  }
}

.header {
  @apply py-2 px-4 relative flex justify-center items-center;
  @apply uppercase leading-none font-bold;
  color: #fff8;
  font-size: 0.7rem;
}

.title {
  text-shadow: -1px -1px #0001;
}

.header-button {
  @apply absolute top-0 right-0 bottom-0 outline-none z-10 px-4;

  &.focus-visible {
    @apply shadow-outline;
  }

  &._toggled {
    @apply bg-gray-900 text-white;
  }

  &:hover {
    @apply text-white;
  }

  &:active {
    @apply bg-gray-800 transition-none text-white;
  }

  & .fa-icon {
    margin: -0.2em 0;
  }
}

.body-wrapper {
  @apply flex flex-grow relative items-stretch;
}

.body {
  @apply grid gap-px text-gray-400 flex-grow;
}

.body > * {
  @apply bg-gray-900;
}

.padded > * {
  @apply p-2;
}

.revealSettings-enter-active,
.revealSettings-leave-active {
  @apply transition duration-300;
}
.revealSettings-enter,
.revealSettings-leave-to {
  @apply opacity-0;
}
</style>
