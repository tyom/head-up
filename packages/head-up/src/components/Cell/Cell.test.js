import { shallowMount } from '@vue/test-utils';
import Cell from './Cell';
import CellSettings from './CellSettings';

const mockProvideNotEditing = {
  isEditing: () => false,
};

const mockProvideEditing = {
  isEditing: () => true,
};

test('default render', () => {
  const wrapper = shallowMount(Cell, {
    provide: mockProvideNotEditing,
  });
  expect(wrapper).toMatchSnapshot();
});

test('with title', () => {
  const wrapper = shallowMount(Cell, {
    propsData: {
      title: 'Important section',
    },
    provide: mockProvideNotEditing,
  });
  expect(wrapper).toMatchSnapshot();
});

test('with content in slots', () => {
  const wrapper = shallowMount(Cell, {
    provide: mockProvideNotEditing,
    slots: {
      default: '<div>1</div><div>2</div><div>3</div>',
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('with content in props', () => {
  const wrapper = shallowMount(Cell, {
    propsData: {
      content: [
        { type: 'CustomComponent', props: { foo: 'custom-prop' } },
        { type: 'CustomComponent', props: { bar: 2 } },
      ],
    },
    stubs: ['CustomComponent'],
    provide: mockProvideNotEditing,
  });
  expect(wrapper).toMatchSnapshot();
});

test('editable', () => {
  const wrapper = shallowMount(Cell, {
    propsData: {
      title: 'Important section',
      editable: true,
    },
    provide: mockProvideEditing,
  });
  expect(wrapper).toMatchSnapshot();
});

test('toggle settings', () => {
  const wrapper = shallowMount(Cell, {
    propsData: {
      title: 'Important section',
      editable: true,
    },
    provide: mockProvideEditing,
  });

  expect(wrapper.find(CellSettings).exists()).toBe(false);
  wrapper.find('.header-button').trigger('click');
  expect(wrapper).toMatchSnapshot();
});
