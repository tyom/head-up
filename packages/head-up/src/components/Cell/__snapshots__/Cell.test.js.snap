// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default render 1`] = `
<div class="Cell">
  <!---->
  <div class="container">
    <div class="body"></div>
    <!---->
  </div>
</div>
`;

exports[`editable 1`] = `
<div class="Cell">
  <header class="header">
    <h2 class="title">Important section</h2> <button class="header-button">
      <v-icon-stub name="ellipsis-h"></v-icon-stub>
    </button>
  </header>
  <div class="container">
    <div class="body"></div>
    <!---->
  </div>
</div>
`;

exports[`toggle settings 1`] = `
<div class="Cell">
  <header class="header">
    <h2 class="title">Important section</h2> <button class="header-button _toggled">
      <v-icon-stub name="ellipsis-h"></v-icon-stub>
    </button>
  </header>
  <div class="container">
    <div class="body"></div>
    <cellsettings-stub name="revealSettings"></cellsettings-stub>
  </div>
</div>
`;

exports[`with content in props 1`] = `
<div class="Cell">
  <!---->
  <div class="container">
    <div class="body u-grid-2-x">
      <customcomponent-stub foo="custom-prop" class="body-item"></customcomponent-stub>
      <customcomponent-stub bar="2" class="body-item"></customcomponent-stub>
    </div>
    <!---->
  </div>
</div>
`;

exports[`with content in slots 1`] = `
<div class="Cell">
  <!---->
  <div class="container">
    <div class="body u-grid-3-x">
      <div>1</div>
      <div>2</div>
      <div>3</div>
    </div>
    <!---->
  </div>
</div>
`;

exports[`with title 1`] = `
<div class="Cell">
  <header class="header">
    <h2 class="title">Important section</h2>
    <!---->
  </header>
  <div class="container">
    <div class="body"></div>
    <!---->
  </div>
</div>
`;
